### PR TITLE
Enable the CopyPropagation pass.

### PR DESCRIPTION
**Pull request by atrick**
For now simply run the pass before SemanticARCOpts. This will probably
be called as a utility from within SemanticARCOpts so it can be
iteratively applied after other ARC-related transformations.

This is rebased on top of the CanonicalOSSA utility:
https://github.com/apple/swift/pull/35241

And depends on a _StringObject.init fix:
https://github.com/apple/swift/pull/35243

### CODE REVIEW

**Comment by atrick** [comment]
@swift-ci test

**Comment by atrick** [comment]
@swift-ci benchmark

**Comment by atrick** [comment]
@swift-ci test source compatibility

**Comment by atrick** [comment]
This passed SCK release testing. ReactiveCocoa fails on main!

This passed regular CI testing. The only lit failure is an AutoDiff check that appeared after rebasing.

File: include/swift/SILOptimizer/Utils/CanonicalOSSALifetime.h [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+// Copyright (c) 2014 - 2020 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+///
+/// Canonicalize the copies and destroys of a single owned or guaranteed OSSA
+/// value.
```
[the remark this answers is not in the document]
Comment (atrick) [reply include/swift/SILOptimizer/Utils/CanonicalOSSALifetime.h:14]: I moved the extensive comments into the header file.

**Comment by atrick** [comment]
@swift-ci test

**Comment by atrick** [comment]
@swift-ci test source compatibility

**Comment by atrick** [comment]
@swift-ci benchmark

**Comment by atrick** [comment]
@swift-ci benchmark

**Comment by atrick** [comment]
The SCK ReactiveCocoa failure is a known issue that's been fixed independently.

**Comment by atrick** [comment]
@swift-ci test

**Comment by atrick** [comment]
@swift-ci test

**Comment by atrick** [comment]
@swift-ci test

**Comment by gottesmm** [comment]
@swift-ci test

**Comment by atrick** [comment]
@swift-ci test

**Comment by atrick** [comment]
@swift-ci test source compatibility

**Comment by atrick** [comment]
@swift-ci benchmark

**Comment by atrick** [comment]
@swift-ci test

**Pull request merged**